consensus/core/src/linearizer.rs

```diff
@@ -0,0 +1,457 @@
+    }
+
+    /// Sort the blocks of the sub-dag by round number. Any deterministic algorithm works.
+    pub fn sort(&mut self) {
+        self.blocks.sort_by_key(|x| x.round());
+    }
+}
+
+impl Display for CommittedSubDag {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "CommittedSubDag(leader={}, index={}, blocks=[",
+            self.leader.digest, self.commit_index
+        )?;
+        for (idx, block) in self.blocks.iter().enumerate() {
+            if idx > 0 {
+                write!(f, ", ")?;
+            }
+            write!(f, "{}", block.digest())?;
+        }
+        write!(f, "])")
+    }
+}
+
+impl fmt::Debug for CommittedSubDag {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}@{}(", self.leader, self.commit_index)?;
+        for block in &self.blocks {
+            write!(f, "{}, ", block.reference())?;
+        }
+        write!(f, ")")
+    }
+}
+
+/// Expand a committed sequence of leader into a sequence of sub-dags.
+#[allow(unused)]
+pub struct Linearizer {
+    // Persistent storage for committed blocks.
+    block_store: Arc<dyn Store>,
+    /// Keep track of all committed blocks to avoid committing the same block twice.
+    committed: HashSet<BlockRef>,
+    /// Keep track of the index of last linearized commit
+    last_commit_index: u64,
+}
+
+#[allow(unused)]
+impl Linearizer {
+    pub fn new(block_store: Arc<dyn Store>) -> Self {
+        Self {
+            block_store,
+            committed: Default::default(),
+            last_commit_index: Default::default(),
+        }
+    }
+
+    pub fn recover_state(&mut self, recovered_state: &CommitObserverRecoveredState) {
+        assert!(self.committed.is_empty());
+        assert_eq!(self.last_commit_index, 0);
+        for commit in recovered_state.commits.iter() {
+            assert!(commit.index > self.last_commit_index);
+            self.last_commit_index = commit.index;
+
+            for block in commit.blocks.iter() {
+                self.committed.insert(*block);
+            }
+            // Leader must be part of the subdag and hence should have been inserted in the loop above.
+            assert!(self.committed.contains(&commit.leader));
+        }
+    }
+
+    /// Collect the sub-dag from a specific anchor excluding any duplicates or blocks that
+    /// have already been committed (within previous sub-dags).
+    fn collect_sub_dag(&mut self, leader_block: VerifiedBlock) -> CommittedSubDag {
+        let mut to_commit = Vec::new();
+
+        let timestamp_ms = leader_block.timestamp_ms();
+        let leader_block_ref = leader_block.reference();
+        let mut buffer = vec![leader_block];
+        assert!(self.committed.insert(leader_block_ref));
+        while let Some(x) = buffer.pop() {
+            to_commit.push(x.clone());
+            let ancestors = self
+                .block_store
+                .read_blocks(x.ancestors())
```

Comment:
It looks to me that we might want to use the `DagState` here since I would expect all the required information to be already cached in memory so we avoid going to the `block_store`. For sure we need to guarantee that no clean up in DagState takes place before the linearisation. If for some reason `DagState` can't be used then I would suggest doing necessary refactorings to ensure that as going to store seems expensive to me.